evennia/contrib/actions/handlers.py

```diff
@@ -0,0 +1,752 @@
+            self.owner.db.actions['active'] = True
+            process_queue(self.owner)
+        else:
+            self.stop(ongoing=True, queued=False)
+            self.owner.db.actions['active'] = False
+
+
+    def add(self, action):
+        """
+        add the action to the bottom of the owner's actions queue
+        """
+        self.owner.msg("|madding {0} to action queue|n".format(action['key']))
+        self.list.append(action)
+
+
+    def pop(self, action):
+        """
+        bring the action up from the queue to the room's action list
+        (assumes it's already been ensured that no other actions from the 
+        owner sharing similar bodyparts are in the room's action list)
+
+        Note: In turn-based mode, popping must only occur while the popped
+        action's owner is referenced by the actions_turnof variable
+        """
+        self.remove(action)
+        action = dict(action)
+        self.owner.location.actions.add(action)
+
+
+    def remove(self, action):
+        """
+        remove the action from the owner's actions queue
+        """
+        self.owner.msg("|mremoving {0} from action queue|n".format(action['key']))
+        self.list.remove(action)
+ 
+
+    def done(self):
+        """
+        During turn-based mode, concludes the turn for the owner
+
+        [WIP] Prohibit "done" from working when the character has issued a
+        movement action, crafting action or any other action with a sufficiently
+        large duration this turn, in order to prevent abuse of turn-based mode
+        for performing large numbers of commands very quickly.
+        """
+        mode = self.owner.location.actions.mode
+        turnof = self.owner.location.actions.turnof
+        if mode == "TB" and turnof == self.owner:
+            # Bring the action system forward
+            self.owner.location.actions.run_script()
+        elif mode == "TB":
+            self.owner.msg("You cannot finish your turn without it being " +
+                "your turn to begin with.")
+        else:
+            self.owner.msg("You cannot finish your turn without being in " +
+                "turn-based mode to begin with.")
+
+    def stop(self, ongoing=True, queued=False):
+        """
+        Cancels the character's ongoing actions if ongoing is set to True,
+        and removes the character's queued actions if queued is set to True.
+        """
+        if queued:
+            # First stop all enqueued actions if the action calls for it, 
+            # to ensure that cancelling the ongoing actions will not load
+            # any actions from the queue
+            for action in self.owner.actions.list:
+                self.owner.actions.remove(action)
+
+        if ongoing:
+            # Then cancel all ongoing actions
+            for action in self.owner.location.actions.list:
+                if action['owner'] == self.owner:
+                    action['at_cancel'](action)
+
+
+    def unpuppet(self):
```

Comment:
Is this to be called specifically when unpuppeting a character? If so this should be mentioned in the docstring. Otherwise this name clashes with the normal meaning of "unpuppet" in Evennia.